=head

# PURPOSE: identify the set of genes considered differentially expressed with age, after taking into consideration (a) correction for multiple testing, (b) high levels of stochastic sampling in single-cell datasets (i.e. % of samples with zero TPM), and (c) whether we are using exact ages or averaged ages per oocyte.

# OUTPUT:

/data/home/Stephen/oocyte_atlas/results/min_depth3000000.min_length70.max_pct_mt10.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 11490 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.35161009573542e-06
/data/home/Stephen/oocyte_atlas/results/min_depth3000000.min_length70.max_pct_mt10.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 11251 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.44404941782953e-06

/data/home/Stephen/oocyte_atlas/results/min_depth3000000.min_length70.max_pct_mt25.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 9334 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 5.35676023141204e-06
/data/home/Stephen/oocyte_atlas/results/min_depth3000000.min_length70.max_pct_mt25.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 9544 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 5.23889354568315e-06

/data/home/Stephen/oocyte_atlas/results/min_depth3000000.min_length70.max_pct_mt100.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 6330 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 7.89889415481833e-06
/data/home/Stephen/oocyte_atlas/results/min_depth3000000.min_length70.max_pct_mt100.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 7355 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 6.79809653297077e-06

/data/home/Stephen/oocyte_atlas/results/min_depth6000000.min_length70.max_pct_mt10.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 11694 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.27569693860099e-06
/data/home/Stephen/oocyte_atlas/results/min_depth6000000.min_length70.max_pct_mt10.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 11340 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.40917107583774e-06

/data/home/Stephen/oocyte_atlas/results/min_depth6000000.min_length70.max_pct_mt25.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 11245 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.44642063139173e-06
/data/home/Stephen/oocyte_atlas/results/min_depth6000000.min_length70.max_pct_mt25.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 10851 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.60787024237398e-06

/data/home/Stephen/oocyte_atlas/results/min_depth6000000.min_length70.max_pct_mt100.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 5749 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 8.69716472429988e-06
/data/home/Stephen/oocyte_atlas/results/min_depth6000000.min_length70.max_pct_mt100.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 7835 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 6.38162093171666e-06

/data/home/Stephen/oocyte_atlas/results/min_depth10000000.min_length70.max_pct_mt10.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 11557 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.32638227913818e-06
/data/home/Stephen/oocyte_atlas/results/min_depth10000000.min_length70.max_pct_mt10.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 11297 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.4259537930424e-06

/data/home/Stephen/oocyte_atlas/results/min_depth10000000.min_length70.max_pct_mt25.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 11209 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.46070122223214e-06
/data/home/Stephen/oocyte_atlas/results/min_depth10000000.min_length70.max_pct_mt25.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 10900 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 4.58715596330275e-06

/data/home/Stephen/oocyte_atlas/results/min_depth10000000.min_length70.max_pct_mt100.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt: there are 6285 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 7.95544948289578e-06
/data/home/Stephen/oocyte_atlas/results/min_depth10000000.min_length70.max_pct_mt100.maturationBoth/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt: there are 8286 genes where > 80% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is 6.03427468018344e-06

=cut

use strict;
use warnings;
use Capture::Tiny qw(capture);

# REQUIREMENTS
my $max_pct_mt = 100; # 10; # 25; # 100;
my $max_seeds  = 25; # should be identical to that used in 6.run_kb_downsampled_oocyte.pl and 7.make_downsampled_tpm_matrix.pl
my $min_depth  = 3000000; # 3000000; # 6000000; # 10000000; # should be identical to that used in 6.run_kb_downsampled_oocyte.pl and 7.make_downsampled_tpm_matrix.pl
my $min_length = 70; # should be identical to that used in 6.run_kb_downsampled_oocyte.pl and 7.make_downsampled_tpm_matrix.pl
my $vivo_vitro = 'Both'; # 'Vivo'; # 'Vitro';
my $params 	   = "min_depth$min_depth.min_length$min_length.max_pct_mt$max_pct_mt.maturation$vivo_vitro";
my $in_file1   = "/data/home/Stephen/oocyte_atlas/results/$params/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AGES.txt"; # from 8a.calculate_spearmans_correls_for_tpm_vs_age_using_R_MII_only.pl
my $in_file2   = "/data/home/Stephen/oocyte_atlas/results/$params/correlations_of_MII_tpm_with_age.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt"; # from 8a.calculate_spearmans_correls_for_tpm_vs_age_using_R_MII_only.pl
if (!(-e($in_file1))) { print "ERROR: cannot find $in_file1\n"; exit 1; }
if (!(-e($in_file2))) { print "ERROR: cannot find $in_file2\n"; exit 1; }

# PARAMETERS
## these are parameters for considering whether a gene is significantly differentially expressed
my $nonzero_threshold 	 = 80;
my $abs_rho_threshold 	 = 0.3;
my $min_mean_over_median = 0.8;
my $max_mean_over_median = 1.2;

# OUTPUT
my $out_file1 = "/data/home/Stephen/oocyte_atlas/results/$params/correlations_of_MII_tpm_with_age_with_adj_p.median_downsampled.tpm.filtered.EXACT_AGES.txt";
my $out_file2 = "/data/home/Stephen/oocyte_atlas/results/$params/correlations_of_MII_tpm_with_age_with_adj_p.median_downsampled.tpm.filtered.EXACT_AND_AVG_AGES.txt";
my $out_file3 = "/data/home/Stephen/oocyte_atlas/results/$params/correlations_of_MII_tpm_with_age.summary.txt";
my $out_file4 = "/data/home/Stephen/oocyte_atlas/results/$params/correlations_of_MII_tpm_with_age.summary_without_bonferroni_correction.txt";
open(OUT1,'>',$out_file1) or die $!; open(OUT2,'>',$out_file2) or die $!; open(OUT3,'>',$out_file3) or die $!; open(OUT4,'>',$out_file4) or die $!;
print OUT3 "Gene name\tEnsembl gene ID\tGene type\tDescription\trho (TPM vs. exact age)\tBonferroni adjusted p-value\trho (TPM vs. exact or average age)\tBonferroni adjusted p-value\tIs the expression of this gene significantly correlated with age irrespective of whether exact-only or exact-or-average ages were used?\n";
print OUT4 "Gene name\tEnsembl gene ID\tGene type\tDescription\trho (TPM vs. exact age)\tuncorrected p-value\trho (TPM vs. exact or average age)\tuncorrected p-value\tIs the expression of this gene significantly correlated with age irrespective of whether exact-only or exact-or-average ages were used?\n";

# STORE SIGNIFICANT CORRELATIONS OF TPM WITH AGE, AS WELL AS UPDATE THE CORRELATION FILES WITH ADJUSTED P-VALUES AND A FLAG TO DENOTE SIGNIFICANCE
my %correlations_p = (); my %correlations_adj_p = (); my %gene_annotations = ();
for(my $x=0;$x<=1;$x++)
	{ my $in_file = ''; my $age = '';
	  if 	($x == 0) { $in_file = $in_file1; $age = 'exact'; 		 }
	  elsif ($x == 1) { $in_file = $in_file2; $age = 'exact_or_avg'; }
	  my %genes = ();
	  open(IN,$in_file) or die $!;
	  while(<IN>)
		{ next if ($. == 1);
		  my $line = $_; chomp($line);
		  my @line = split(/\t/,$line);
		  my $gene_id = $line[0]; my $pct_non_zero = $line[6];
		  next if ($pct_non_zero < $nonzero_threshold); # CHECKPOINT: exclude genes where too few samples have non-zero expression (which indicates too high a level of stochastic sampling)
		  $genes{$gene_id}++;
		}
	  close(IN) or die $!;
	  my $num_genes = scalar keys %genes;
	  my $bonferroni_threshold = 0.05/$num_genes;
	  if 	($x == 0) { print OUT1 "Gene ID\tGene name\tGene type\tDescription\tNo. of MII oocytes\tNo. of MII oocytes with non-zero expression\t% of MII oocytes with non-zero expression\tAge range (years)\tMin TPM - max TPM\tTPM range\tMean TPM\tMedian TPM\tMean TPM/median TPM\trho\tp-value\tBonferroni corrected p-value\tIs expression level considered significantly differentially expressed with age? (criteria: absolute rho > $abs_rho_threshold, mean/median TPM > $min_mean_over_median and < $max_mean_over_median, and adj. p-value < $bonferroni_threshold)\n"; }
	  elsif ($x == 1) { print OUT2 "Gene ID\tGene name\tGene type\tDescription\tNo. of MII oocytes\tNo. of MII oocytes with non-zero expression\t% of MII oocytes with non-zero expression\tAge range (years)\tMin TPM - max TPM\tTPM range\tMean TPM\tMedian TPM\tMean TPM/median TPM\trho\tp-value\tBonferroni corrected p-value\tIs expression level considered significantly differentially expressed with age? (criteria: absolute rho > $abs_rho_threshold, mean/median TPM > $min_mean_over_median and < $max_mean_over_median, and adj. p-value < $bonferroni_threshold)\n"; }
	  print "$in_file: there are $num_genes genes where > $nonzero_threshold% of the oocytes have a non-zero TPM, so the Bonferroni adjusted-p threshold is $bonferroni_threshold\n";
	  open(IN,$in_file) or die $!;
	  while(<IN>)
		{ next if ($. == 1);
		  my $line = $_; chomp($line);
		  my @line = split(/\t/,$line);
		  my $gene_id = $line[0]; my $gene_name = $line[1]; my $gene_type = $line[2]; my $gene_desc = $line[3]; my $num_oocytes = $line[4]; my $num_non_zero = $line[5]; my $pct_non_zero = $line[6]; my $age_range = $line[7]; my $min_max_tpm = $line[8]; my $tpm_range = $line[9]; my $mean_tpm = $line[10]; my $median_tpm = $line[11]; my $mean_over_median_tpm = $line[12]; my $rho = $line[13]; my $p = $line[14];
		  $gene_annotations{$gene_name}{gene_id}   = $gene_id;
		  $gene_annotations{$gene_name}{gene_type} = $gene_type;
		  $gene_annotations{$gene_name}{gene_desc} = $gene_desc;
		  my $adj_p = $p*$num_genes; if ($adj_p > 1) { $adj_p = 1; }
		  
		  my $is_signif_p = 'no'; my $is_signif_adj_p = 'no';
		  if (($pct_non_zero > $nonzero_threshold) and (abs($rho) > $abs_rho_threshold) and ($p < 0.05) and ($mean_over_median_tpm > $min_mean_over_median) and ($mean_over_median_tpm < $max_mean_over_median))
			{ $is_signif_p = 'yes'; }
		  if (($pct_non_zero > $nonzero_threshold) and (abs($rho) > $abs_rho_threshold) and ($adj_p < 0.05) and ($mean_over_median_tpm > $min_mean_over_median) and ($mean_over_median_tpm < $max_mean_over_median))
			{ $is_signif_adj_p = 'yes'; }
		  
		  if    ($x == 0) { print OUT1 "$gene_id\t$gene_name\t$gene_type\t$gene_desc\t$num_oocytes\t$num_non_zero\t$pct_non_zero\t$age_range\t$min_max_tpm\t$tpm_range\t$mean_tpm\t$median_tpm\t$mean_over_median_tpm\t$rho\t$p\t$adj_p\t$is_signif_adj_p\n"; }
		  elsif ($x == 1) { print OUT2 "$gene_id\t$gene_name\t$gene_type\t$gene_desc\t$num_oocytes\t$num_non_zero\t$pct_non_zero\t$age_range\t$min_max_tpm\t$tpm_range\t$mean_tpm\t$median_tpm\t$mean_over_median_tpm\t$rho\t$p\t$adj_p\t$is_signif_adj_p\n"; }
		  
		  if ($is_signif_adj_p eq 'yes') # genes which have significant correlations after correction for multiple testing
			{ $correlations_adj_p{$gene_name}{$age}{rho}   = $rho;
			  $correlations_adj_p{$gene_name}{$age}{adj_p} = $adj_p;
			}
		  if ($is_signif_p eq 'yes') # genes which have significant correlations WITH NO correction for multiple testing
			{ $correlations_p{$gene_name}{$age}{rho} = $rho;
			  $correlations_p{$gene_name}{$age}{p}   = $p;
			}
		}
	  close(IN) or die $!;
	}
close(OUT1) or die $!; close(OUT2) or die $!;

# OUTPUT LIST OF GENES WHOSE EXPRESSION LEVEL SIGNIFICANTLY CORRELATES WITH AGE (BONFERRONI-CORRECTED P < 0.05)
my @gene_names = ();
while((my $gene_name,my $irrel)=each(%correlations_adj_p))
	{ push(@gene_names,$gene_name); }
my @sorted_gene_names = sort {$a cmp $b} @gene_names;
foreach my $gene_name (@sorted_gene_names)
	{ my $exact_rho = 'NA'; my $exact_p = 'NA';
	  if (exists($correlations_adj_p{$gene_name}{'exact'}))
		{ $exact_rho = $correlations_adj_p{$gene_name}{'exact'}{rho};
		  $exact_p   = $correlations_adj_p{$gene_name}{'exact'}{adj_p};
		}
	  my $averg_rho = 'NA'; my $averg_p = 'NA';
	  if (exists($correlations_adj_p{$gene_name}{'exact_or_avg'}))
		{ $averg_rho = $correlations_adj_p{$gene_name}{'exact_or_avg'}{rho};
		  $averg_p   = $correlations_adj_p{$gene_name}{'exact_or_avg'}{adj_p};
		}
	  my $signif_in_both = 'no';
	  if (($exact_rho !~ /^NA$/) and ($averg_rho !~ /^NA$/)) { $signif_in_both = 'yes'; }
	  print OUT3 "$gene_name\t$gene_annotations{$gene_name}{gene_id}\t$gene_annotations{$gene_name}{gene_type}\t$gene_annotations{$gene_name}{gene_desc}\t$exact_rho\t$exact_p\t$averg_rho\t$averg_p\t$signif_in_both\n";
	}
close(OUT3) or die $!;

# OUTPUT LIST OF GENES WHOSE EXPRESSION LEVEL SIGNIFICANTLY CORRELATES WITH AGE (P < 0.05; NO CORRECTION FOR MULTIPLE TESTING)
@gene_names = ();
while((my $gene_name,my $irrel)=each(%correlations_p))
	{ push(@gene_names,$gene_name); }
@sorted_gene_names = sort {$a cmp $b} @gene_names;
foreach my $gene_name (@sorted_gene_names)
	{ my $exact_rho = 'NA'; my $exact_p = 'NA';
	  if (exists($correlations_p{$gene_name}{'exact'}))
		{ $exact_rho = $correlations_p{$gene_name}{'exact'}{rho};
		  $exact_p   = $correlations_p{$gene_name}{'exact'}{p};
		}
	  my $averg_rho = 'NA'; my $averg_p = 'NA';
	  if (exists($correlations_p{$gene_name}{'exact_or_avg'}))
		{ $averg_rho = $correlations_p{$gene_name}{'exact_or_avg'}{rho};
		  $averg_p   = $correlations_p{$gene_name}{'exact_or_avg'}{p};
		}
	  my $signif_in_both = 'no';
	  if (($exact_rho !~ /^NA$/) and ($averg_rho !~ /^NA$/)) { $signif_in_both = 'yes'; }
	  print OUT4 "$gene_name\t$gene_annotations{$gene_name}{gene_id}\t$gene_annotations{$gene_name}{gene_type}\t$gene_annotations{$gene_name}{gene_desc}\t$exact_rho\t$exact_p\t$averg_rho\t$averg_p\t$signif_in_both\n";
	}
close(OUT4) or die $!;
exit 1;